webclient/pid/client.py

```diff
@@ -0,0 +1,107 @@
+import logging
+
+import requests
+from django.conf import settings
+from django.utils.translation import gettext as _
+from dokument.models import Dokument
+from lokalita.models import Lokalita
+from pas.models import SamostatnyNalez
+from pid.model_serializers import DokumentSerializer, LokalitaSerializer, SamostatnyNalezSerializer
+from requests.auth import HTTPBasicAuth
+
+from webclient.settings.base import DATACITE_URL, DATACITE_USER, DATACITE_USER_PASSWORD
+
+logger = logging.getLogger(__name__)
+
+
+class DoiWriteError(Exception):
+    pass
+
+
+class DoiHasBeenAlreadyPublishedError(DoiWriteError):
+    pass
+
+
+class DoiHasNotBeenPublishedYetError(DoiWriteError):
+    pass
+
+
+class DoiNoTransactionError(DoiWriteError):
+    pass
+
+
+class DigitalObjectIdentifierClient:
+    auth = HTTPBasicAuth(DATACITE_USER, DATACITE_USER_PASSWORD)
+    headers = {"Content-Type": "application/vnd.api+json"}
+
+    def __init__(self, record):
+        self.record = record
+        if isinstance(self.record, Dokument):
+            self.serializer = DokumentSerializer(self.record)
+            self.attribute_name = "doi"
+        elif isinstance(self.record, Lokalita):
+            self.serializer = LokalitaSerializer(self.record)
+            self.attribute_name = "igsn"
+        elif isinstance(self.record, SamostatnyNalez):
+            self.serializer = SamostatnyNalezSerializer(self.record)
+            self.attribute_name = "igsn"
+        else:
+            logger.error("doi.client.DigitalObjectIdentifierClient.invalid_record_class")
+            raise ValueError(_("doi.client.DigitalObjectIdentifierClient.invalid_record_class"))
+
+    def _check_response_status(self, response):
+        if not str(response.status_code).startswith("2"):
+            logger.error(
+                "doi.client.DigitalObjectIdentifierClient._check_response_status.error",
+                extra={
+                    "ident_cely": self.serializer.get_ident_cely(),
+                    "status_code": response.status_code,
+                },
+            )
+            raise DoiWriteError
+
+    def _get_record_url(self):
+        return f"{DATACITE_URL.rstrip('/')}/{settings.DOI_PREFIX}/{self.serializer.get_ident_cely()}"
+
+    def check_record_exists(self):
+        response = requests.get(self._get_record_url(), auth=self.auth)
+        return str(response.status_code).startswith("2")
```

Comment:
Co kdyz to jen vrati 500? Nebylo by lepsi checkovat 200 vs 404 vs ostatni?

Reply:
Upravil jsem.